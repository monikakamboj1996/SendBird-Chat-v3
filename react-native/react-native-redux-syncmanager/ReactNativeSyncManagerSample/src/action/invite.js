
const Action = {
	Type: {
		USER_FETCH: 'user-fetch'
	}
};
Action.fetchUsers = users => {
	return {
		type: Action.Type.USER_FETCH,
		payload: { users }
	};
};

export default Action;