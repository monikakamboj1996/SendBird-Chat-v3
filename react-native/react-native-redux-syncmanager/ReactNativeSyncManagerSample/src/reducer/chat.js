import Action from '../action/chat';
import { findMessageIndex, getMessageIndex } from '../utils';

const _initialState = {
  channel: null,
  messages: [],
  inputMode: 'chat'
};

export const reducer = (state = _initialState, action) => {
  const newState = { ...state };
  switch (action.type) {
    case Action.Type.MESSAGE_SELECT: {
      newState.messages = [...newState.messages];
      for (let i = 0; i < newState.messages.length; i++) {
        newState.messages[i]._selected = newState.messages[i].isIdentical(action.payload.message);
      }
      break;
    }
    case Action.Type.MESSAGE_DESELECT_ALL: {
      newState.messages = [...newState.messages];
      for (let i = 0; i < newState.messages.length; i++) {
        newState.messages[i]._selected = false;
      }
      break;
    }
    case Action.Type.MESSAGE_ADD: {
      newState.messages = [...newState.messages];
      for (let i = 0; i < action.payload.messages.length; i++) {
        const message = action.payload.messages[i];
        const isFailed = message.messageId === 0 && message.requestState === 'failed';
        const index = findMessageIndex(message, newState.messages, isFailed);
        if (index >= 0) {
          newState.messages.splice(index, 0, message);
          if (index > 0) {
            newState.messages[index - 1]._isPreviousMessageSentBySameUser =
              newState.messages[index - 1].sender &&
              newState.messages[index].sender &&
              newState.messages[index - 1].sender.userId === newState.messages[index].sender.userId;
          }
          if (index < newState.messages.length - 1) {
            newState.messages[index]._isPreviousMessageSentBySameUser =
              newState.messages[index].sender &&
              newState.messages[index + 1].sender &&
              newState.messages[index].sender.userId === newState.messages[index + 1].sender.userId;
          }
        }
      }
      break;
    }
    case Action.Type.MESSAGE_UPDATE: {
      newState.messages = [...newState.messages];
      for (let i = 0; i < action.payload.messages.length; i++) {
        const message = action.payload.messages[i];
        const index = getMessageIndex(message, newState.messages);
        if (index >= 0) {
          message._isPreviousMessageSentBySameUser = newState.messages[index]._isPreviousMessageSentBySameUser;
          newState.messages[index] = message;
          if (action.payload.isTransformFailedToManual) {
            newState.messages[index]._isManualResend = true;
          }
        }
      }
      break;
    }
    case Action.Type.MESSAGE_REMOVE: {
      newState.messages = [...newState.messages];
      for (let i = 0; i < action.payload.messages.length; i++) {
        const message = action.payload.messages[i];
        const index = getMessageIndex(message, newState.messages);
        if (index >= 0) {
          newState.messages.splice(index, 1);
          if (index < newState.messages.length - 1) {
            newState.messages[index]._isPreviousMessageSentBySameUser =
              newState.messages[index + 1].sender &&
              newState.messages[index].sender &&
              newState.messages[index + 1].sender.userId === newState.messages[index].sender.userId;
          }
        }
      }
      break;
    }
    case Action.Type.MESSAGE_CLEAR: {
      newState.messages = [];
      break;
    }
  }
  return newState;
};
