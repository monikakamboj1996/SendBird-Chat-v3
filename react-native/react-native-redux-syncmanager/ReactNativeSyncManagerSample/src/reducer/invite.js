import Action from "../action/invite";

const _initialState = {
	users: [],
	err: null
};

export const reducer = (state = _initialState, action) => {
	const newState = { ...state };
	switch(action.type) {
		case Action.Type.USER_FETCH: {
			newState.users = newState.users.concat(action.payload.users);
			break;
		}
	}
	return newState;
};