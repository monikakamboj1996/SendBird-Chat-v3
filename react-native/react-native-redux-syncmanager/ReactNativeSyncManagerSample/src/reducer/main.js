import Action from '../action/main';
import { findChannelIndex, getChannelIndex } from '../utils';

const _initialState = {
  channels: []
};

export const reducer = (state = _initialState, action) => {
  const newState = { ...state };
  switch (action.type) {
    case Action.Type.CHANNEL_ADD: {
      newState.channels = [...newState.channels];
      for (let i = 0; i < action.payload.channels.length; i++) {
        const channel = action.payload.channels[i];
        const index = findChannelIndex(channel, newState.channels);
        if (index >= 0) {
          newState.channels.splice(index, 0, channel);
        }
      }
      break;
    }
    case Action.Type.CHANNEL_UPDATE: {
      newState.channels = [...newState.channels];
      for (let i = 0; i < action.payload.channels.length; i++) {
        const channel = action.payload.channels[i];
        const index = getChannelIndex(channel, newState.channels);
        if (index >= 0) {
          newState.channels[index] = channel;
        }
      }
      break;
    }
    case Action.Type.CHANNEL_REMOVE: {
      newState.channels = [...newState.channels];
      for (let i = 0; i < action.payload.channels.length; i++) {
        const channel = action.payload.channels[i];
        const index = getChannelIndex(channel, newState.channels);
        if (index >= 0) {
          newState.channels.splice(index, 1);
        }
      }
      break;
    }
    case Action.Type.CHANNEL_MOVE: {
      newState.channels = [...newState.channels];
      for (let i = 0; i < action.payload.channels.length; i++) {
        const channel = action.payload.channels[i];
        const from = getChannelIndex(channel, newState.channels);
        if (from >= 0) {
          newState.channels.splice(from, 1);
        }
        const to = findChannelIndex(channel, newState.channels);
        newState.channels.splice(to, 0, channel);
      }
      break;
    }
    case Action.Type.CHANNEL_CLEAR: {
      newState.channels = [];
      break;
    }
  }
  return newState;
};
