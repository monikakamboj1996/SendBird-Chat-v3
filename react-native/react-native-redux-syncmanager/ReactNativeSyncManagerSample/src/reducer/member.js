
import Action from "../action/member";

const _initialState = {
	members: []
};

export const reducer = (state = _initialState, action) => {
	const newState = { ...state };
	switch(action.type) {
		case Action.Type.MEMBER_INIT: {
			newState.members = action.payload.members;
			break;
		}
		case Action.Type.MEMBER_UPDATE: {
			newState.members = [ ...action.payload.members ];
			break;
		}
	}
	return newState;
};