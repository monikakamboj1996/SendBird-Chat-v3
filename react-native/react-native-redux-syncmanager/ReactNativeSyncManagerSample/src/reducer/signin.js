
import Action from "../action/signin";

const _initialState = {
	authenticating: false,
	user: null,
	err: null,
	typedUserId: '',
	typedNickname: ''
};

export const reducer = (state = _initialState, action) => {
	const newState = { ...state };
	switch(action.type) {
		case Action.Type.AUTHENTICATE: {
			newState.authenticating = true;
			break;
		}
		case Action.Type.SIGNIN_SUCCESS: {
			newState.authenticating = false;
			newState.user = action.payload.user;
			newState.err = null;
			break;
		}
		case Action.Type.SIGNIN_FAILED: {
			newState.authenticating = false;
			newState.user = null;
			newState.err = action.payload.err;
			break;
		}
	}
	return newState;
};