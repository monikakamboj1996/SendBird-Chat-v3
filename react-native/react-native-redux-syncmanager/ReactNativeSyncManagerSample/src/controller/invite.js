import React from 'react';
import { View } from 'react-native';
import { Button } from 'react-native-elements';
import { connect } from 'react-redux';
import Icon from 'react-native-vector-icons/Ionicons';
import { FlatList } from 'react-native-gesture-handler';

import SendBird from 'sendbird';

import { style } from '../style/invite';
import Action from '../action/invite';
import UserView from '../view/user';
import { toast } from '../utils';

let _selectedUserIds = [];

export class InviteController extends React.Component {
  static navigationOptions = ({ navigation }) => {
    const channel = navigation.getParam('channel', null);
    return {
      title: !channel ? 'Create new channel' : 'Invite',
      headerRight: (
        <View style={{ flexDirection: 'row' }}>
          <Button
            containerStyle={style.headerInviteContainer}
            buttonStyle={style.headerInvite}
            icon={<Icon name="md-add" color="#7d62d9" size={28}></Icon>}
            onPress={() => {
              if (!channel) {
                if (_selectedUserIds.length > 0) {
                  const sb = SendBird.getInstance();
                  const params = new sb.GroupChannelParams();
                  params.addUserIds(_selectedUserIds);
                  sb.GroupChannel.createChannel(params, (channel, err) => {
                    if (!err) {
                      navigation.replace('Chat', { channel });
                    } else {
                      toast('Failed to create channel.');
                    }
                  });
                } else {
                  toast('Choose at least 1 user to invite.');
                }
              } else {
                channel.inviteWithUserIds(_selectedUserIds, (res, err) => {
                  if (!err) {
                    navigation.pop();
                  } else {
                    toast('Failed to invite users.');
                  }
                  isLoading = false;
                });
              }
            }}
          />
        </View>
      )
    };
  };
  constructor(props) {
    super(props);
    const sb = SendBird.getInstance();
    this.query = sb.createUserListQuery();
    this.channel = this.props.navigation.getParam('channel', null);
    this.state = {
      users: []
    };
  }
  static getDerivedStateFromProps(props, state) {
    state.users = props.users;
    return state;
  }
  componentDidMount() {
    this.next();
  }
  componentWillUnmount() {
    _selectedUserIds = [];
  }
  toggle(selected, user) {
    if (selected) {
      if (_selectedUserIds.indexOf(user.userId) < 0) {
        _selectedUserIds.push(user.userId);
      }
    } else {
      const index = _selectedUserIds.indexOf(user.userId);
      if (index >= 0) {
        _selectedUserIds.splice(index, 1);
      }
    }
  }
  next() {
    if (this.query.hasNext) {
      this.query.next((users, err) => {
        if (!err) {
          let filteredUsers = [];
          if (this.channel) {
            for (let i in users) {
              if (this.channel.members.map(m => m.userId).indexOf(users[i].userId) < 0) {
                filteredUsers.push(users[i]);
              }
            }
          } else {
            filteredUsers = users;
          }
          if (filteredUsers.length < this.query.limit && this.query.hasNext) {
            setTimeout(() => this.next(), 200);
          } else {
            this.props.fetchUsers(filteredUsers);
          }
        }
      });
    }
  }
  render() {
    return (
      <View>
        <FlatList
          data={this.state.users}
          keyExtractor={user => user.userId + ''}
          onEndReached={this.next.bind(this)}
          renderItem={bundle => (
            <UserView
              user={bundle.item}
              selectable={true}
              selected={_selectedUserIds.indexOf(bundle.item.userId) >= 0}
              onPress={this.toggle.bind(this)}
            />
          )}
        />
      </View>
    );
  }
}

export default connect(
  previousState => {
    const { invite } = previousState;
    return { ...invite };
  },
  Action
)(InviteController);
