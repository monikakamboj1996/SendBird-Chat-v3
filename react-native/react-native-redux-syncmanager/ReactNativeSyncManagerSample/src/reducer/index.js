import { combineReducers } from 'redux';
import { reducer as main } from './main';
import { reducer as chat } from './chat';
import { reducer as invite } from './invite';
import { reducer as member } from './member';
import { reducer as signin } from './signin';

export default combineReducers({
  main,
  chat,
  invite,
  member,
  signin
});
