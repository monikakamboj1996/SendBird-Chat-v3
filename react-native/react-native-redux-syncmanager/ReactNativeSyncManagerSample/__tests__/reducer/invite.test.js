import { reducer } from '../../src/reducer/invite';
import Action from '../../src/action/invite';

describe('reducer/invite', () => {
  test('checking initial state', () => {
    const action = { type: 'nonexistent' };
    expect(reducer(undefined, action)).toMatchSnapshot();
  });

  test('USER_FETCH', () => {
    const action = {
      type: Action.Type.USER_FETCH,
      payload: {
        users: ['a', 'b', 'c']
      }
    };
    expect(reducer({ users: ['d', 'e', 'f'], err: null }, action)).toMatchSnapshot();
  });
});
