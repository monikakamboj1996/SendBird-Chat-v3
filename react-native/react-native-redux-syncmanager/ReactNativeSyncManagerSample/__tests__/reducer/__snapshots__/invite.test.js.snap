// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer/invite USER_FETCH 1`] = `
Object {
  "err": null,
  "users": Array [
    "d",
    "e",
    "f",
    "a",
    "b",
    "c",
  ],
}
`;

exports[`reducer/invite checking initial state 1`] = `
Object {
  "err": null,
  "users": Array [],
}
`;
